Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs

```diff
@@ -0,0 +1,679 @@
+                pointerState.OnMove(m_CurrentTime, position, targetDisplay);
+        }
+
+        void OnSubmitPerformed(InputAction.CallbackContext ctx)
+        {
+            DispatchFromCallback(Event.From(new NavigationEvent
+            {
+                type = NavigationEvent.Type.Submit,
+                direction = NavigationEvent.Direction.None,
+                timestamp = m_CurrentTime,
+                eventSource = GetEventSource(ctx),
+                playerId = k_DefaultPlayerId,
+                eventModifiers = m_EventModifiers
+            }));
+        }
+
+        void OnCancelPerformed(InputAction.CallbackContext ctx)
+        {
+            DispatchFromCallback(Event.From(new NavigationEvent
+            {
+                type = NavigationEvent.Type.Cancel,
+                direction = NavigationEvent.Direction.None,
+                timestamp = m_CurrentTime,
+                eventSource = GetEventSource(ctx),
+                playerId = k_DefaultPlayerId,
+                eventModifiers = m_EventModifiers
+            }));
+        }
+
+        void OnClickPerformed(InputAction.CallbackContext ctx, EventSource eventSource, PointerEvent.Button button)
+        {
+            ref var state = ref GetPointerStateForSource(eventSource);
+
+            var asTouchscreenDevice = ctx.control.device is Touchscreen ? (Touchscreen)ctx.control.device : null;
+            var asTouchControl = ctx.control is TouchControl ? (TouchControl)ctx.control : null;
+            var pointerIndex = FindTouchFingerIndex(asTouchscreenDevice, ctx);
+
+            m_ResetSeenEventsOnUpdate = true;
+            if (asTouchControl != null || asTouchscreenDevice != null)
+                m_SeenTouchEvents = true;
+
+            var wasPressed = state.ButtonsState.Get(button);
+            var isPressed = ctx.ReadValueAsButton();
+            state.OnButtonChange(m_CurrentTime, button, wasPressed, isPressed);
+
+            DispatchFromCallback(Event.From(new PointerEvent
+            {
+                type = isPressed ? PointerEvent.Type.ButtonPressed : PointerEvent.Type.ButtonReleased,
+                pointerIndex = pointerIndex,
+                position = state.LastPosition,
+                deltaPosition = Vector2.zero,
+                scroll = Vector2.zero,
+                displayIndex = state.LastDisplayIndex,
+                tilt = Vector2.zero,
+                twist = 0.0f,
+                pressure = 0.0f,
+                isInverted = false,
+                button = button,
+                buttonsState = state.ButtonsState,
+                clickCount = state.ClickCount,
+                timestamp = m_CurrentTime,
+                eventSource = eventSource,
+                playerId = k_DefaultPlayerId,
+                eventModifiers = m_EventModifiers
+            }));
+        }
+
+        void OnLeftClickPerformed(InputAction.CallbackContext ctx) => OnClickPerformed(ctx, GetEventSource(ctx), PointerEvent.Button.MouseLeft);
```

Comment:
Could do GetEventSource call inside the above OnClickPerformed function and avoid passing as parameter?